Source/DafnyCore/Prelude/Sequences.dfy

```diff
@@ -0,0 +1,796 @@
+  //   { Seq#Contains(Seq#Build(s, v), x) }
+  //   Seq#Contains(Seq#Build(s, v), x) <==> (v == x || Seq#Contains(s, x)));
+  lemma {:extract_pattern Contains(Build(s, v), x)} BuildContains(s: Seq, v: Box, x: Box)
+    ensures Contains(Build(s, v), x) <==> v == x || Contains(s, x)
+  {
+    var valList := Cons(v, Nil);
+    var b := s.Append(valList);
+    assert b == Build(s, v);
+
+    assert Ping: v == x ==> Contains(valList, x) by {
+      assert Index(valList, 0) == v;
+    }
+    assert Pong: Contains(valList, x) ==> v == x by {
+      if Contains(valList, x) {
+        var i :| 0 <= i < valList.Length() && Index(valList, i) == x;
+        assert valList.Length() == 1;
+      }
+    }
+
+    calc {
+      Contains(b, x);
+      { AppendContains(s, valList, x); }
+      Contains(s, x) || Contains(valList, x);
+      { reveal Ping, Pong; }
+      v == x || Contains(s, x);
+    }
+  }
+
+  // boogie:
+  // axiom (forall s: Seq, n: int, x: Box ::
+  //   { Seq#Contains(Seq#Take(s, n), x) }
+  //   Seq#Contains(Seq#Take(s, n), x) <==>
+  //     (exists i: int :: { Seq#Index(s, i) }
+  //       0 <= i && i < n && i < Seq#Length(s) && Seq#Index(s, i) == x));
+  lemma {:extract_pattern Contains(Take(s, n), x)} TakeContains(s: Seq, n: int, x: Box)
+    ensures Contains(Take(s, n), x) <==>
+            exists i: int {:extract_pattern Index(s, i)} :: 0 <= i < n && i < Length(s) && Index(s, i) == x
+  {
+    if
+    case n < 0 =>
+      calc {
+        Contains(Take(s, n), x);
+        { assert Take(s, n) == Empty(); }
+        Contains(Empty(), x);
+        false;
+        exists i :: 0 <= i < n && i < Length(s) && Index(s, i) == x;
+      }
+
+    case 0 <= n <= s.Length() =>
+      var (prefix, suffix) := s.Split(n);
+      var t := s.Take(n);
+      assert t == prefix;
+      assert t.Length() == n by {
+        s.LengthTakeDrop(n);
+      }
+      assert L: forall i :: 0 <= i < Length(t) ==> Index(t, i) == Index(s, i) by {
+        forall i | 0 <= i < Length(t)
+          ensures Index(t, i) == Index(s, i)
+        {
+          prefix.AppendAt(suffix, i);
+        }
+      }
+
+      calc {
+        Contains(t, x);
+        exists i :: 0 <= i < Length(t) && Index(t, i) == x;
+        { reveal L; }
+        exists i :: 0 <= i < n && i < Length(s) && Index(s, i) == x;
+      }
+
+    case s.Length() < n =>
+      calc {
+        Contains(Take(s, n), x);
+        Contains(s, x);
+        exists i :: 0 <= i < Length(s) && Index(s, i) == x;
+        { assert Length(s) < n; }
+        exists i :: 0 <= i < n && i < Length(s) && Index(s, i) == x;
+      }
+  }
+
+  // boogie:
+  // axiom (forall s: Seq, n: int, x: Box ::
+  //   { Seq#Contains(Seq#Drop(s, n), x) }
+  //   Seq#Contains(Seq#Drop(s, n), x) <==>
+  //     (exists i: int :: { Seq#Index(s, i) }
+  //       0 <= n && n <= i && i < Seq#Length(s) && Seq#Index(s, i) == x));
+  lemma {:extract_pattern Contains(Drop(s, n), x)} DropContains(s: Seq, n: int, x: Box)
```

Comment:
I'm tempted to have a different name for this attribute, too, though I'm not sure what...

Reply:
I left these as is. If we think of a better naming scheme in the future, we can change them then.